### PR TITLE
Add release command to build-tools CLI

### PR DESCRIPTION
**Pull request by tylerbutler**
This PR implements the `release` command, which is used to ensure that a release branch is in good condition, then walks the user through the release.

The release process itself is described by a state machine. This is a visual representation:

image

The basic flow of the command is:

- The command starts and oclif (our CLI infrastructure) parses flags and arguments, validates them, etc.
- The command's _run_ method is called, which calls an internal _stateLoop_ function.
- That function loops through the machine states, passing each state to the command's state handler function.
- The state handler handles the states it knows about, and if it doesn't handle the state, it passes it to its parent's state handler. Thus, the parent classes are relevant to the overall implementation of the state handling.

The base classes are helpful to share state handling between commands. In a future PR I'll add the `release prep` command which will re-use much of the base class functionality.

## Guidance for reviewers

This is not in a state that I prefer to share PRs. In particular, it's missing a lot of comments, and there are inconsistencies in the APIs that are pretty obvious (still, point them out when you see them!). Thus I recommend focusing on the structure and flow of the code rather than going over functions with a fine-toothed comb.

I would particular like feedback on:

- How the code is organized - both at the file level and at the module level.
- It is a goal that given the name of a state and its code, it should be easy to understand what it's doing and the conditions under which it succeeds or fails. Did I succeed? The big picture is better understood using the state machine diagram.
- There are also some half-baked experiments in some places, especially around testing and type-safety. I think I went overboard. Please save me! 😄

### CODE REVIEW

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts [hunk 1]
```diff
@@ -161,7 +161,7 @@ export class GitRepo {
      * Fetch Tags
      */
     public async fetchTags() {
-        return await this.exec(`fetch --tags`, `fetch tags`);
+        return await this.exec(`fetch --tags --force`, `fetch tags`);
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:164]: Is this change needed? Or was it just for testing?

File: build-tools/packages/version-tools/src/commands/version.ts [hunk 2]
```diff
@@ -9,13 +9,15 @@ import { table } from "table";
...
     bumpInternalVersion,
-    DEFAULT_PUBLIC_VERSION,
     fromInternalScheme,
     toInternalScheme,
     getVersionRange,
     isInternalVersionScheme,
 } from "../internalVersionScheme";
 
+/** The default public version used to generate Fluid internal versions when none is provided. */
+const DEFAULT_PUBLIC_VERSION = "2.0.0";
```
Comment (Josmithr) [inline build-tools/packages/version-tools/src/commands/version.ts:19]: What is the long-term plan for this default? I'm always nervous when I see something like this codified.

**Comment by Josmithr** [comment]
I really like the new structure. The individual pieces are way easier to follow now. 

The only comment I have about the larger organization is that I think it would be nice to have the state machine definitions a bit closer to the command types that use them. Bouncing back and forth between the two locations was a little rough. That said, I know you're also intending for more code-reuse, so maybe that separation will prove more useful in the future. I still think this is a HUGE improvement, so I would take this as a nitpicky suggestion over a real issue.

**Comment by tylerbutler** [comment]
> The only comment I have about the larger organization is that I think it would be nice to have the state machine definitions a bit closer to the command types that use them. Bouncing back and forth between the two locations was a little rough.

You know, it was the way you described before and I split the definitions out. I think my reasoning was that it was easier to describe what was the state machine definition vs the command that uses an instance of the state machine. That said, I think maybe splitting the base commands into different files is a better approach. Right now all the base classes are in one file -- I can make that a module instead.

File: build-tools/packages/version-tools/src/commands/version.ts (hunk 2 above)
Comment (tylerbutler) [reply build-tools/packages/version-tools/src/commands/version.ts:19]: Good question. I figured that since it's a default it would remain "for all time". But I could just remove the default. It makes the command mildly more annoying to use, but it's probably the right tradeoff.

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts (hunk 1 above)
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:164]: I consistently have to force because of the archive/0.39 tag. I've been meaning to ask @curtisman if maybe we could just delete that tag.

**Comment by tylerbutler** [comment]
Replaced by #11706.
